File: Common Algorithm/Graph Algorithms/Depth First Search/dfs.py
from collections import defaultdict

class Graph:
    def __init__(self, v):
        self.graph = defaultdict(list)
        self.visited = [False] * v

    def addEdge(self, u, v):
        self.graph[u].append(v)

    def dfsUtil(self, root, visited):
        visited[root] = True
        print(root, end = " ")

        for i in self.graph[root]:
            if not self.visited[i]:
                self.dfsUtil(i, visited)

    def dfs(self, root):
        self.dfsUtil(root, self.visited)

g = Graph(8)

g.addEdge(0, 1)
g.addEdge(0, 3)
g.addEdge(0, 4)
g.addEdge(3, 2)
g.addEdge(2, 5)
g.addEdge(2, 7)
g.addEdge(7, 6)
g.addEdge(6, 1)

g.dfs(0)

File: Common Algorithm/Graph Algorithms/Depth First Search/dfs.js
class Graph {
  constructor(v) {
    this.adj = new Array(v);
    // for (let i = 0; i < v; i++) this.adj[i] = [];
    this.adj = Array.from({ length: v }, () => []);
    this.visited = new Array(v);
    this.visited.fill(false);
  }

  addEdge(v, w) {
    this.adj[v].push(w);
  }

  DFSUtil(root) {
    this.visited[root] = true;
    console.log(root);

    for (let adjElem of this.adj[root]) {
      if (!this.visited[adjElem]) this.DFSUtil(adjElem);
    }
  }

  DFS(root) {
    this.DFSUtil(root);
  }
}

g = new Graph(8);

g.addEdge(0, 1);
g.addEdge(0, 3);
g.addEdge(0, 4);
g.addEdge(3, 2);
g.addEdge(2, 5);
g.addEdge(2, 7);
g.addEdge(7, 6);
g.addEdge(6, 1);

g.DFS(0);
